### PR TITLE
Autodoc for smart contracts

### LINKED ISSUE #2350: Documentation for smart contracts public API
It's very likely that some stakers and community members will not interact with our smart contracts via the CLI or the staking UIs, and use directly the smart contracts API. While [this page from the docs](https://docs.nucypher.com/en/latest/architecture/contracts.html) offers a nice overview, we need proper documentation of the main methods, perhaps fetching them directly from our current natspec comments.

### PR DESCRIPTION
**Pull request by vzotova**
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [X] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [ ] 2
- [X] 3

**What this does:**
> High-level idea of the changes introduced in this PR.  

Generates documentation for smart contracts from natspec

> List relevant API changes (if any), as well as related PRs and issues.

None

**Issues fixed/closed:**

Fixes #2350

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

Improves docs. 
We already have descriptions inside contract sources but now it's much more convenient to show/explain how API works

**Notes for reviewers:**
> What should reviewers focus on?
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?
- New resulting htmls for contracts in build artifacts
- `generate_doc.py` - natspec -> rst script
- Changes in `setup.py` for using git urls in requirements

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -98,7 +100,21 @@ def run(self):
 
 def read_requirements(path):
     with open(os.path.join(BASE_DIR, path)) as f:
-        _pipenv_flags, *requirements = f.read().split('\n')
+        _pipenv_flags, *lines = f.read().split('\n')
+
+    # Transforms VCS requirements to PEP 508
```
Comment (KPrasch) [inline setup.py:105]: :exploding_head: - A real testament to the current state of python dependency management.

File: docs-requirements.txt [hunk 2]
```diff
@@ -9,7 +9,9 @@ idna==2.10
 imagesize==1.2.0
 incremental==17.5.0
 jinja2==2.11.2
+git+git://github.com/inspirehep/jsonschema2rst@d211d9709046431a6b6a4594c97c22d8713d854b#egg=jsonschema2rst
```
Comment (cygnusv) [inline docs-requirements.txt:12]: Why this specific commit? I guess there's no package in pypi?

Comment (KPrasch) [reply docs-requirements.txt:12]: The package maintainer does not adhere to a convention versioning scheme on PyPI, so this is the only way to reference the previous version (current one has a bug) .

File: scripts/solidity_doc/generate_doc.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+def merge_and_update(source, destination):
+    """Update values in source dictionary and merge it to the destination dictionary"""
+
+    for key, value in source.items():
+        if isinstance(value, dict):
+            destination[key] = merge_and_update(value, destination.get(key, {}))
+        else:
+            # | symbol causes warnings in docs compilation
+            # doubles ` symbol to make it as code piece in rst
+            value = value.replace('`', '``').replace('|', '') if isinstance(value, str) else value
```
Comment (KPrasch) [inline scripts/solidity_doc/generate_doc.py:55]: I'd like to see either Issue or TODOs for further discussion and evolution.

File: scripts/solidity_doc/generate_doc.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+             'NuCypherToken',
+             'StakingInterface',
+             'StakingInterfaceRouter',
+             'AbstractStakingContract',
+             'InitializableStakingContract',
+             'PoolingStakingContract',
+             'WorkLockPoolingContract']
+
+
+def merge_and_update(source, destination):
```
Comment (KPrasch) [inline scripts/solidity_doc/generate_doc.py:46]: Type hints please! :-)

File: scripts/solidity_doc/generate_doc.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    GlobalLoggerSettings.start_console_logging()
+
+    bundle = SourceBundle(base_path=SOLIDITY_SOURCE_ROOT)
+    contracts = multiversion_compile(source_bundles=[bundle])
+
+    base_path = BASE_DIR / 'docs' / 'source' / 'contracts_api'
+    base_path.mkdir(exist_ok=True)
+
+    patch()
```
Comment (KPrasch) [inline scripts/solidity_doc/generate_doc.py:71]: Another opportunity for a TODO and future improvement.

**Review by KPrasch — APPROVED** [review]
This is a stellar effort I'm happy to see you've broken ground on this.  It's such a good Idea I think it can one day grow to be it's own library, or perhaps a utility into eth-brownie.

It really is amazing to see the contract docs rendered onto RTD via CI :tada:

File: docs-requirements.txt (hunk 2 above)
Comment (vzotova) [reply docs-requirements.txt:12]: New version (0.1.4) includes too old pytest in normal installation, old version (0.1.0) has issues with metadata

File: .circleci/config.yml [hunk 6]
```diff
@@ -679,7 +679,7 @@ jobs:
       - prepare_environment
       - run:
           name: Install Documentation Build Dependencies
-          command: pip3 install --use-feature=2020-resolver --user .[docs]
```
[the remark this answers is not in the document]
Comment (vzotova) [reply .circleci/config.yml:682]: 2020 resolver is default with current version of CircleCi's pip

File: scripts/solidity_doc/generate_doc.py (hunk 3 above)
Comment (vzotova) [reply scripts/solidity_doc/generate_doc.py:56]: ✔️

File: scripts/solidity_doc/generate_doc.py (hunk 4 above)
Comment (vzotova) [reply scripts/solidity_doc/generate_doc.py]: ✔️

File: scripts/solidity_doc/generate_doc.py (hunk 5 above)
Comment (vzotova) [reply scripts/solidity_doc/generate_doc.py:72]: ✔️

File: scripts/solidity_doc/generate_doc.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+def generate_doc() -> None:
+    """Compile solidity contracts, extract json docs and generate rst files from them"""
+
+    GlobalLoggerSettings.start_console_logging()
+
+    bundle = SourceBundle(base_path=SOLIDITY_SOURCE_ROOT)
+    contracts = multiversion_compile(source_bundles=[bundle])
+
+    base_path = BASE_DIR / 'docs' / 'source' / 'contracts_api'
```
Comment (derekpierre) [inline scripts/solidity_doc/generate_doc.py:69]: 😍

**Review by derekpierre — COMMENTED** [review]
🎸 - haven't tried it out locally since updating solc takes forever on a mac, but I can see the build version - https://115681-100446516-gh.circle-artifacts.com/0/docs/build/html/index.html

Should we think about having one API topn section and then having two collapsable subsections, one for `nucypher` and the other for contracts? Right now, they are not collapsable and both of them together make our docs look bulky:

Screen Shot 2020-12-24 at 10 51 27 AM

**Comment by vzotova** [comment]
> Should we think about having one API overall section and then having two collapsable subsections, one for `nucypher` and the other for contracts? Right now, they are not collapsable and both of them together make our docs look bulky:

All of these contracts are nucypher contracts. But of course we can make some subcategories. Waiting for suggestions

**Comment by cygnusv** [comment]
> > Should we think about having one API overall section and then having two collapsable subsections, one for `nucypher` and the other for contracts? Right now, they are not collapsable and both of them together make our docs look bulky:
> 
> All of these contracts are nucypher contracts. But of course we can make some subcategories. Waiting for suggestions

First thoughts on categories:
* Main contracts: The big four + WorkLock
* Upgradeability and proxies: Dispatcher, StakingInterface, StakingInterfaceRouter, Upgradeable   (not really sure about this category name)
* Staking and Pooling Contracts

**Comment by cygnusv** [comment]
BTW, the `TokenRecipient` contract maybe shouldn't appear as documentation? It's only a minor interface.

**Comment by vzotova** [comment]
> BTW, the `TokenRecipient` contract maybe shouldn't appear as documentation? It's only a minor interface.

anybody can use with token contract and some their contract which inherits TokenRecipient

**Comment by vzotova** [comment]
> First thoughts on categories:
> 
> * Main contracts: The big four + WorkLock
> * Upgradeability and proxies: Dispatcher, StakingInterface, StakingInterfaceRouter, Upgradeable   (not really sure about this category name)
> * Staking and Pooling Contracts

added subcategories https://115837-100446516-gh.circle-artifacts.com/0/docs/build/html/contracts_api/index.html

File: setup.py (hunk 1 above)
Comment (KPrasch) [reply setup.py:105]: I must admit this patching is too hacky to love. Ideally this code can be removed once there is a solution to the dependency issue.  Let's include a TODO at minimum.

Comment (vzotova) [reply setup.py:106]: ✔️

**Review by cygnusv — APPROVED** [review]

**Review by derekpierre — APPROVED** [review]
🎸 - looks good and approved.

We could potentially collapse it further (maybe in a separate PR (?)) by combining with the existing API section. Something like:

```
API
    - Platform
        - nucypher.blockchain
        - nucypher.characters
        - nucypher.config
        - ...
    - Contracts
        - Token Contracts
        - Main Contracts
        - Upgradeability and proxies
        - Staking and Pooling Contracts
```

@vzotova , @cygnusv , @KPrasch wdyt?

**Comment by cygnusv** [comment]
@derekpierre I agree. The docs need a slight reorganization in general, but that's outside the scope is this PR, IMO.

**Pull request merged**